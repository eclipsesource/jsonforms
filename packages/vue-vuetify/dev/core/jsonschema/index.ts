import type { JsonSchema } from '@jsonforms/core';
import rule from './specification/rule.json';
import schema from './specification/schema.json';
import uischema from './specification/uischema.json';

export const jsonSchemaDraft7 = {
  uri: 'http://json-schema.org/draft-07/schema',
  schema: schema as any as JsonSchema,
};

export const uiSchema = {
  uri: 'http://jsonforms.io/uischema',
  schema: uischema as JsonSchema,
};
export const ruleSchema = {
  uri: 'http://jsonforms.io/uischema/rule',
  schema: rule as JsonSchema,
};
