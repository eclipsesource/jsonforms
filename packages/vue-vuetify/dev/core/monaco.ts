import * as monaco from 'monaco-editor';

import editorWorker from 'monaco-editor/esm/vs/editor/editor.worker?worker';
import jsonWorker from 'monaco-editor/esm/vs/language/json/json.worker?worker';

self.MonacoEnvironment = {
  getWorker(_, label) {
    if (label === 'json') {
      return new jsonWorker();
    }
    return new editorWorker();
  },
};

export default monaco;
export type MonacoApi = typeof monaco;
