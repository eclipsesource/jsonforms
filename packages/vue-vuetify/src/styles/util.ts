import cloneDeep from 'lodash/cloneDeep';
import mergeWith from 'lodash/mergeWith';
import type { Styles } from './styles';

export const classes = (
  strings: TemplateStringsArray,
  ...variables: any[]
): string => {
  return strings
    .reduce((acc, curr, index) => {
      return `${acc}${curr}${variables[index] || ''}`;
    }, '')
    .trim();
};

/**
 * Helper function to merge two styles definitions. The contained classes will be combined, not overwritten.
 *
 * Example usage:
 * ```ts
 * const myStyles = mergeStyles(defaultStyles, { control: { root: 'mycontrol' } });
 * ```
 */
export const mergeStyles = (
  stylesA: Partial<Styles>,
  stylesB: Partial<Styles>,
): Partial<Styles> => {
  const styles = cloneDeep(stylesA);
  mergeWith(styles, stylesB, (aValue, bValue) => {
    if (typeof aValue === 'string' && typeof bValue === 'string') {
      return `${aValue} ${bValue}`;
    }
    return undefined;
  });
  return styles;
};
