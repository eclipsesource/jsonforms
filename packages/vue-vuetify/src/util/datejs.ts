import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import localeData from 'dayjs/plugin/localeData';
import localizedFormat from 'dayjs/plugin/localizedFormat';
import timezone from 'dayjs/plugin/timezone'; // dependent on utc plugin
import utc from 'dayjs/plugin/utc';

import type { MaskTokens, MaskType } from 'maska';

// required for the custom save formats in the date, time and date-time pickers
dayjs.extend(customParseFormat);
dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(localeData);
dayjs.extend(localizedFormat);

export const expandLocaleFormat = (format: string): string => {
  return format.replace(/\b(LT|LTS|L{1,4})\b/g, function (match, p1) {
    // Replace each captured value with a specific value
    switch (p1) {
      case 'LT':
      case 'LTS':
      case 'L':
      case 'LL':
      case 'LLL':
      case 'LLLL':
        return dayjs.localeData().longDateFormat(p1);
      default:
        return match; // Return the match unchanged if not in the set
    }
  });
};

export const parseDateTime = (
  data: string | number | dayjs.Dayjs | Date | null | undefined,
  format: string | string[] | undefined,
): dayjs.Dayjs | null => {
  if (!data) {
    return null;
  }
  const dayjsData = dayjs(data, format);
  if (!dayjsData.isValid()) {
    return null;
  }
  return dayjsData;
};

export const convertDayjsToMaskaFormat = (
  dayjsFormat: string,
): { mask: MaskType; tokens: MaskTokens } => {
  function* nextKey(preservedKeys: string): Generator<string> {
    let currentCharCode = 'A'.charCodeAt(0);

    while (currentCharCode <= 65535) {
      // Unicode character range
      const currentChar = String.fromCharCode(currentCharCode);
      if (!preservedKeys.includes(currentChar)) {
        yield currentChar;
      }
      currentCharCode++;
    }
  }

  function RegExpLiteral(str: string) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }

  class Tokens {
    private tokens: MaskTokens;
    private charGenerator;

    constructor(preservedKeys: string) {
      this.tokens = {};
      this.charGenerator = nextKey(preservedKeys);
    }

    public getTokens(): MaskTokens {
      return this.tokens;
    }

    public getTokenKey(token: MaskTokens[keyof MaskTokens]): string {
      const entry = Object.entries(this.tokens).find(
        (entry) =>
          entry[1].pattern.toString() === token.pattern.toString() &&
          entry[1].optional === token.optional,
      );
      if (entry) {
        return entry[0];
      }

      throw new Error(
        'Key for pattern ' + token.pattern.toString() + ' not found.',
      );
    }

    public registerToken(token: MaskTokens[keyof MaskTokens]): string {
      const entry = Object.entries(this.tokens).find(
        (entry) =>
          entry[1].pattern.toString() === token.pattern.toString() &&
          entry[1].optional === token.optional,
      );
      if (entry) {
        return entry[0];
      }

      const key = this.charGenerator.next().value;

      this.tokens[key] = token;
      return key;
    }
  }

  // defintions from - https://day.js.org/docs/en/parse/string-format#list-of-all-available-parsing-tokens
  const dayjsTokens = [
    'YYYY', // Four-digit year - example: 2001
    'YY', //Two-digit year - example: 01
    'M', // Month, beginning at 1 - example: 1-12
    'MM', // Month, 2-digits - example: 01-12
    'MMM', // The abbreviated month name - example: Jan-Dec
    'MMMM', // The full month name - example: January-December
    'D', // Day of month - example: 1-31
    'DD', // Day of month, 2-digits - example: 01-31
    'H', // Hours - example: 0-23
    'HH', // Hours, 2-digits - example: 00-23
    'h', // Hours, 12-hour clock - example: 1-12
    'hh', // Hours, 12-hour clock, 2-digits - example: 01-12
    'm', // Minutes - example: 0-59
    'mm', // Minutes, 2-digits - example: 00-59
    's', // Seconds - example: 0-59
    'ss', // Seconds, 2-digits - example: 00-59
    'S', // Hundreds of milliseconds, 1-digit - example: 0-9
    'SS', // Tens of milliseconds, 2-digits - example: 00-99
    'SSS', // Milliseconds, 3-digits - example: 000-999
    'Z', // Offset from UTC - example: -05:00
    'ZZ', // Compact offset from UTC, 2-digits - example: -0500
    'A', // Post or ante meridiem, upper-case - example: AM PM
    'a', // Post or ante meridiem, lower-case - example: am pm,
    'X', // Unix timestamp - example 1410715640.579,
    'x', // Unix ms timestamp - example 1410715640579
  ].sort((a, b) => b.length - a.length);

  const parts = dayjsFormat
    .split(new RegExp(`(${dayjsTokens.join('|')})`))
    .filter(Boolean);

  const reservedChars = parts
    .filter((part) => !dayjsTokens.includes(part))
    .join('');
  const tokens = new Tokens(reservedChars);

  // need to prebuild all keys to tokens are are going to be used in tokenFunction
  tokens.registerToken({ pattern: /[0-9]/ });
  tokens.registerToken({ pattern: /1/ });
  tokens.registerToken({ pattern: /[0-2]/ });
  tokens.registerToken({ pattern: /[0-2]/, optional: true });
  tokens.registerToken({ pattern: /[0-1]/ });
  tokens.registerToken({ pattern: /[1-9]/ });

  let escapedMonths = dayjs.monthsShort().map((month) => RegExpLiteral(month));
  let regexPattern = `\\b(${escapedMonths.join('|')})\\b`;
  tokens.registerToken({ pattern: new RegExp(regexPattern, 'i') });

  escapedMonths = dayjs.months().map((month) => RegExpLiteral(month));
  regexPattern = `\\b(${escapedMonths.join('|')})\\b`;

  tokens.registerToken({ pattern: new RegExp(regexPattern, 'i') });
  tokens.registerToken({ pattern: /[1-3]/ });
  tokens.registerToken({ pattern: /[0-1]/, optional: true });
  tokens.registerToken({ pattern: /[0-9]/, optional: true });
  tokens.registerToken({ pattern: /[0-3]/ });
  tokens.registerToken({ pattern: /[0-3]/, optional: true });
  tokens.registerToken({ pattern: /[0-5]/ });
  tokens.registerToken({ pattern: /[ap]/ });
  tokens.registerToken({ pattern: /m/ });
  tokens.registerToken({ pattern: /[AP]/ });
  tokens.registerToken({ pattern: /M/ });
  tokens.registerToken({ pattern: /[+-]/ });
  tokens.registerToken({ pattern: /[0-4]/ });
  tokens.registerToken({ pattern: /:/ });
  tokens.registerToken({ pattern: /[.]/ });
  tokens.registerToken({ pattern: /[.]/, optional: true });

  const tokenFunction = (value: string): string => {
    let index = 0;

    const result: (string | MaskTokens[keyof MaskTokens])[] = [];
    for (const part of parts) {
      if (!part || part === '') {
        continue;
      }
      if (index > value.length) {
        break;
      }

      if (part === 'YYYY') {
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        index += 4;
      } else if (part === 'YY' || part == 'SS') {
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        index += 2;
      } else if (part === 'M') {
        result.push({ pattern: /1/ });
        if (value.charAt(index) === '1') {
          if (/[0-2]/.test(value.charAt(index + 1))) {
            result.push({ pattern: /[0-2]/ });
            index += 1;
          } else if (value.charAt(index + 1) === '') {
            result.push({ pattern: /[0-2]/, optional: true });
          }
        }
        index += 1;
      } else if (part === 'MM' || part == 'hh') {
        result.push({ pattern: /[0-1]/ });
        result.push(
          value.charAt(index) === '0'
            ? { pattern: /[1-9]/ }
            : { pattern: /[0-2]/ },
        );
        index += 2;
      } else if (part === 'MMM') {
        const escapedMonths = dayjs
          .monthsShort()
          .map((month) => RegExpLiteral(month));
        const regexPattern = `\\b(${escapedMonths.join('|')})\\b`;

        result.push({ pattern: new RegExp(regexPattern, 'i') });
        let monthSpecified = false;

        for (const month of dayjs.monthsShort()) {
          if (
            index < value.length &&
            value.substring(index).startsWith(month)
          ) {
            index += month.length;
            monthSpecified = true;
            break;
          }
        }
        if (!monthSpecified) {
          // can't continue until the months is not matched
          break;
        }
      } else if (part === 'MMMM') {
        const escapedMonths = dayjs
          .months()
          .map((month) => RegExpLiteral(month));
        const regexPattern = `\\b(${escapedMonths.join('|')})\\b`;

        result.push({ pattern: new RegExp(regexPattern, 'i') });
        let monthSpecified = false;

        for (const month of dayjs.months()) {
          if (
            index < value.length &&
            value.substring(index).startsWith(month)
          ) {
            index += month.length;
            monthSpecified = true;
            break;
          }
        }
        if (!monthSpecified) {
          // can't continue until the months is not matched
          break;
        }
      } else if (part === 'D') {
        result.push({ pattern: /[1-9]/ });
        if (/[1-3]/.test(value.charAt(index))) {
          if (value.charAt(index) === '3') {
            if (/[0-1]/.test(value.charAt(index + 1))) {
              result.push({ pattern: /[0-1]/ });
              index += 1;
            } else if (value.charAt(index + 1) === '') {
              result.push({ pattern: /[0-1]/, optional: true });
            }
          } else {
            if (/\d/.test(value.charAt(index + 1))) {
              result.push({ pattern: /[0-9]/ });
              index += 1;
            } else if (value.charAt(index + 1) === '') {
              result.push({ pattern: /[0-9]/, optional: true });
            }
          }
        }
        index += 1;
      } else if (part === 'DD') {
        result.push({ pattern: /[0-3]/ });
        if (value.charAt(index) === '3') {
          result.push({ pattern: /[0-1]/ });
        } else if (value.charAt(index) === '0') {
          result.push({ pattern: /[1-9]/ });
        } else {
          result.push({ pattern: /[0-9]/ });
        }
        index += 2;
      } else if (part == 'H') {
        result.push({ pattern: /[0-9]/ });
        if (value.charAt(index) === '2') {
          if (/[0-3]/.test(value.charAt(index + 1))) {
            result.push({ pattern: /[0-3]/ });
            index += 1;
          } else if (value.charAt(index + 1) === '') {
            result.push({ pattern: /[0-3]/, optional: true });
          }
        } else if (value.charAt(index) === '1') {
          if (/\d/.test(value.charAt(index + 1))) {
            result.push({ pattern: /[0-9]/ });
            index += 1;
          } else if (value.charAt(index + 1) === '') {
            result.push({ pattern: /[0-9]/, optional: true });
          }
        }
        index += 1;
      } else if (part == 'HH') {
        result.push({ pattern: /[0-2]/ });
        if (/[0-1]/.test(value.charAt(index))) {
          result.push({ pattern: /[0-9]/ });
        } else if (value.charAt(index) === '2') {
          result.push({ pattern: /[0-3]/ });
        }
        index += 2;
      } else if (part == 'h') {
        result.push({ pattern: /[1-9]/ });
        if (value.charAt(index) === '1') {
          if (/[0-2]/.test(value.charAt(index + 1))) {
            result.push({ pattern: /[0-2]/ });
            index += 1;
          } else if (value.charAt(index + 1) === '') {
            result.push({ pattern: /[0-2]/, optional: true });
          }
        }
        index += 1;
      } else if (part == 'm' || part == 's') {
        result.push({ pattern: /[0-9]/ });
        if (/[1-5]/.test(value.charAt(index))) {
          if (/\d/.test(value.charAt(index + 1))) {
            result.push({ pattern: /[0-9]/ });
            index += 1;
          } else if (value.charAt(index + 1) === '') {
            result.push({ pattern: /[0-9]/, optional: true });
          }
        }
        index += 1;
      } else if (part == 'mm' || part == 'ss') {
        result.push({ pattern: /[0-5]/ });
        result.push({ pattern: /[0-9]/ });
        index += 2;
      } else if (part == 'a') {
        result.push({ pattern: /[ap]/ });
        result.push({ pattern: /m/ });
        index += 2;
      } else if (part == 'A') {
        result.push({ pattern: /[AP]/ });
        result.push({ pattern: /M/ });
        index += 2;
      } else if (part == 'Z' || part == 'ZZ') {
        //GMT-12 to GMT+14
        result.push({ pattern: /[+-]/ });
        result.push({ pattern: /[0-1]/ });

        if (value.charAt(index + 1) === '0') {
          result.push({ pattern: /[0-9]/ });
        } else if (value.charAt(index + 1) === '1') {
          result.push(
            value.charAt(index) === '+'
              ? { pattern: /[0-4]/ }
              : { pattern: /[0-2]/ },
          );
        }
        if (part === 'Z') {
          result.push({ pattern: /:/ });
          index += 1;
        }
        result.push({ pattern: /[0-5]/ });
        result.push({ pattern: /[0-9]/ });
        index += 5;
      } else if (part == 'S') {
        result.push({ pattern: /[0-9]/ });
        index += 1;
      } else if (part == 'SSS') {
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        result.push({ pattern: /[0-9]/ });
        index += 3;
      } else if (part == 'X') {
        // number of digits 13
        const times = 13;
        for (let i = 0; i < times; i++) {
          result.push({ pattern: /[0-9]/ });
        }
        index += times;
      } else if (part == 'x') {
        // number of digits 10 for seconds
        const times = 10;
        for (let i = 0; i < times; i++) {
          result.push({ pattern: /[0-9]/ });
        }

        if (value.charAt(index + times) === '.') {
          result.push({ pattern: /[.]/ });
          index += 1;

          // check for optional 3 digits after .
          for (let i = 0; i < 3; i++) {
            if (/\d/.test(value.charAt(index + times + 1 + i))) {
              result.push({ pattern: /[0-9]/ });
              index += 1;
            } else if (value.charAt(index + times + 1 + i) === '') {
              result.push({ pattern: /[0-9]/, optional: true });
              break;
            }
          }
        } else if (value.charAt(index + times) === '') {
          result.push({ pattern: /[.]/, optional: true });
        }

        index += times;
      } else {
        result.push(part);
        index += part.length;
      }
    }

    const mask = result
      .map((part) =>
        typeof part === 'string' ? part : tokens.getTokenKey(part),
      )
      .join('');

    return mask;
  };

  return { mask: tokenFunction, tokens: tokens.getTokens() };
};
