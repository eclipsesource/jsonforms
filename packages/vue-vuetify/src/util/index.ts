export * from './composition';
export * from './datejs';
export * from './options';
export * from './validator';
