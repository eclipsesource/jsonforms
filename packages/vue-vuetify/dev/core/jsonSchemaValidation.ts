import type { JsonSchema } from '@jsonforms/core';
import monaco, { type MonacoApi } from './monaco';

import isEqual from 'lodash/isEqual';
import { jsonSchemaDraft7, ruleSchema, uiSchema } from '../core/jsonschema';

/**
 * Register a new schema for the Json language, if it isn't already registered.
 * Schemas are identified by their uri and fileMatch rule, so that they don't
 * leak into unrelated Json editors.
 * @param editor
 *  The monaco editor
 * @param schemas
 *  Schemas to register
 */
export const addSchema = (
  editor: MonacoApi,
  schemas: {
    uri: string;
    fileMatch?: string[];
    schema?: JsonSchema;
  }[],
): void => {
  const registeredSchemas =
    editor.languages.json.jsonDefaults.diagnosticsOptions.schemas;
  if (registeredSchemas === undefined || registeredSchemas.length === 0) {
    editor.languages.json.jsonDefaults.setDiagnosticsOptions({
      validate: true,
      allowComments: false,
      enableSchemaRequest: false,
      schemaRequest: 'warning',
      schemaValidation: 'error',
      schemas: [...schemas],
    });
  } else {
    for (const schema of schemas) {
      const fileMatch = schema.fileMatch;

      const existingSchemaIndex = registeredSchemas.findIndex(
        (registeredSchema) =>
          isEqual(registeredSchema.fileMatch, fileMatch) &&
          isEqual(registeredSchema.uri, schema.uri),
      );
      if (existingSchemaIndex !== -1) {
        registeredSchemas[existingSchemaIndex] = { ...schema };
      } else {
        registeredSchemas.push({ ...schema });
      }
    }

    editor.languages.json.jsonDefaults.setDiagnosticsOptions({
      validate: true,
      allowComments: false,
      enableSchemaRequest: false,
      schemaRequest: 'warning',
      schemaValidation: 'error',
      schemas: [...registeredSchemas],
    });
  }
};

/**
 * Configures the Monaco Editor to validate the input against JSON Schema Draft 7.
 */
export const configureJsonSchemaValidation = (
  editor: MonacoApi,
  fileMatch: string[],
): void => {
  /** Note that the Monaco Editor only supports JSON Schema Draft 7 itself,
   * so if we also want to support a later standard we still have to formalize
   * it in JSON Schema Draft 7*/
  addSchema(editor, [{ ...jsonSchemaDraft7, fileMatch }]);
};

/**
 * Configures the Monaco Editor to validate the input against the UI Schema meta-schema.
 */
export const configureUISchemaValidation = (
  editor: MonacoApi,
  fileMatch: string[],
): void => {
  /** Note that the Monaco Editor only supports JSON Schema Draft 7 itself,
   * so if we also want to support a later standard we still have to formalize
   * it in JSON Schema Draft 7*/
  addSchema(editor, [
    { ...jsonSchemaDraft7 },
    { ...ruleSchema },
    { ...uiSchema, fileMatch },
  ]);
};

/**
 * Configures the Monaco Editor to validate the input against JSON Schema model schema.
 */
export const configureDataValidation = (
  editor: MonacoApi,
  uri: string,
  fileMatch: string,
  schema: JsonSchema,
): void => {
  /** Note that the Monaco Editor only supports JSON Schema Draft 7 itself,
   * so if we also want to support a later standard we still have to formalize
   * it in JSON Schema Draft 7*/
  addSchema(editor, [{ schema, uri, fileMatch: [fileMatch] }]);
};

export const getMonacoModelForUri = (
  modelUri: monaco.Uri,
  initialValue: string | undefined,
): monaco.editor.ITextModel => {
  const value = initialValue ?? '';
  let model = monaco.editor.getModel(modelUri);
  if (model) {
    model.setValue(value);
  } else {
    model = monaco.editor.createModel(value, 'json', modelUri);
  }
  return model;
};
