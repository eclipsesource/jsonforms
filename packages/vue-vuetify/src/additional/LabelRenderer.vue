<template>
  <v-label
    v-if="label.visible"
    :class="styles.label.root"
    v-bind="vuetifyProps('v-label')"
  >
    {{ label.text }}
  </v-label>
</template>

<script lang="ts">
import {
  rankWith,
  uiTypeIs,
  type JsonFormsRendererRegistryEntry,
  type LabelElement,
} from '@jsonforms/core';
import {
  rendererProps,
  useJsonFormsLabel,
  type RendererProps,
} from '@jsonforms/vue';
import { defineComponent } from 'vue';
import { VLabel } from 'vuetify/components';
import { useVuetifyLabel } from '../util';

const labelRenderer = defineComponent({
  name: 'label-renderer',
  components: {
    VLabel,
  },
  props: {
    ...rendererProps<LabelElement>(),
  },
  setup(props: RendererProps<LabelElement>) {
    return useVuetifyLabel(useJsonFormsLabel(props));
  },
});

export default labelRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: labelRenderer,
  tester: rankWith(1, uiTypeIs('Label')),
};
</script>
