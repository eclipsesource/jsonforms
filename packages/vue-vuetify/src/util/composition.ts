import { aliases as faIcons } from '@/icons/fa';
import type { IconAliases } from '@/icons/icons';
import { aliases as mdiIcons } from '@/icons/mdi';
import {
  Resolve,
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  type ControlElement,
  type DispatchPropsOfControl,
  type JsonFormsSubStates,
  type JsonSchema,
  type UISchemaElement,
} from '@jsonforms/core';
import type Ajv from 'ajv';
import type { ErrorObject } from 'ajv';
import cloneDeep from 'lodash/cloneDeep';
import debounce from 'lodash/debounce';
import get from 'lodash/get';
import isPlainObject from 'lodash/isPlainObject';
import merge from 'lodash/merge';
import {
  computed,
  inject,
  provide,
  ref,
  type ComputedRef,
  type InjectionKey,
} from 'vue';
import type { IconOptions } from 'vuetify';
import { useStyles } from '../styles';

export const IconSymbol: InjectionKey<Required<IconOptions>> =
  Symbol.for('vuetify:icons');

export const useControlAppliedOptions = <
  T extends { config: any; uischema: UISchemaElement },
  I extends {
    control: ComputedRef<T>;
  },
>(
  input: I,
) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options),
    ),
  );
};

export const useLayoutAppliedOptions = <
  T extends { config: any; uischema: UISchemaElement },
  I extends {
    layout: ComputedRef<T>;
  },
>(
  input: I,
) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options),
    ),
  );
};

export const useComputedLabel = <
  T extends { label: string; required: boolean },
  I extends { control: ComputedRef<T> },
>(
  input: I,
  appliedOptions: ReturnType<typeof useControlAppliedOptions>,
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk,
    );
  });
};

/**
 * Adds styles, appliedOptions and vuetifyProps
 */
export const useVuetifyLabel = <
  T extends {
    uischema: UISchemaElement;
    config: any;
  },
  I extends {
    label: ComputedRef<T>;
  },
>(
  input: I,
) => {
  const styles = useStyles(input.label.value.uischema);
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.label.value.config),
      cloneDeep(input.label.value.uischema.options),
    ),
  );
  const vuetifyProps = (path: string) => {
    const props = get(appliedOptions.value?.vuetify, path);

    return props && isPlainObject(props) ? props : {};
  };
  return {
    ...input,
    appliedOptions,
    vuetifyProps,
    styles,
  };
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  T extends {
    uischema: ControlElement;
    path: string;
    config: any;
    label: string;
    description: string;
    required: boolean;
    errors: string;
    id: string;
    visible: boolean;
  },
  I extends {
    control: ComputedRef<T>;
  } & DispatchPropsOfControl,
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v,
  debounceWait?: number,
) => {
  const touched = ref(false);

  const changeEmitter =
    typeof debounceWait === 'number'
      ? debounce(input.handleChange, debounceWait)
      : input.handleChange;

  const onChange = (value: any) => {
    changeEmitter(input.control.value.path, adaptValue(value));
  };

  const appliedOptions = useControlAppliedOptions(input);
  const isFocused = ref(false);

  const handleFocus = () => {
    isFocused.value = true;
  };

  const handleBlur = () => {
    touched.value = true;
    isFocused.value = false;
  };

  const filteredErrors = computed(() => {
    return touched.value || !appliedOptions.value.enableFilterErrorsBeforeTouch
      ? input.control.value.errors
      : '';
  });

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription,
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  const vuetifyProps = (path: string) => {
    const props = get(appliedOptions.value?.vuetify, path);

    return props && isPlainObject(props) ? props : {};
  };

  const overwrittenControl = computed(() => {
    return {
      ...input.control.value,
      errors: filteredErrors.value,
    };
  });

  const rawErrors = computed(() => input.control.value.errors);

  return {
    ...input,
    control: overwrittenControl,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    vuetifyProps,
    persistentHint,
    computedLabel,
    touched,
    handleBlur,
    handleFocus,
    rawErrors,
  };
};

export const useJsonForms = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms couldn't be injected. Are you within JSON Forms?",
    );
  }

  return jsonforms;
};

export const useTranslator = () => {
  const jsonforms = useJsonForms();

  if (!jsonforms.i18n || !jsonforms.i18n.translate) {
    throw new Error(
      "'jsonforms i18n couldn't be injected. Are you within JSON Forms?",
    );
  }

  const translate = computed(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return jsonforms.i18n!.translate!;
  });

  return translate;
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <
  T extends { config: any; uischema: UISchemaElement },
  I extends { layout: ComputedRef<T> },
>(
  input: I,
) => {
  const appliedOptions = useLayoutAppliedOptions(input);

  const vuetifyProps = (path: string) => {
    const props = get(appliedOptions.value?.vuetify, path);

    return props && isPlainObject(props) ? props : {};
  };

  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
    vuetifyProps,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <
  T extends {
    label: string;
    required: boolean;
    config: any;
    uischema: UISchemaElement;
    schema: JsonSchema;
    data: any;
    childErrors: ErrorObject[];
  },
  I extends {
    control: ComputedRef<T>;
  },
>(
  input: I,
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const vuetifyProps = (path: string) => {
    const props = get(appliedOptions.value?.vuetify, path);

    return props && isPlainObject(props) ? props : {};
  };

  const childLabelForIndex = (index: number | null) => {
    if (index === null) {
      return '';
    }
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp),
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  const filteredChildErrors = computed(() => {
    // supress childErrors unless touch filtering is disabled
    // otherwise all child errors will show, irrespective of their control touch state
    const filtered: ErrorObject[] = appliedOptions.value
      ?.enableFilterErrorsBeforeTouch
      ? []
      : input.control.value.childErrors;
    return filtered;
  });
  const overwrittenControl = computed(() => {
    return {
      ...input.control.value,
      childErrors: filteredChildErrors.value,
    };
  });
  return {
    ...input,
    control: overwrittenControl,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
    vuetifyProps,
    rawChildErrors: input.control.value.childErrors,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <
  T extends { config: any; uischema: UISchemaElement },
  I extends {
    control: ComputedRef<T>;
  },
>(
  input: I,
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const vuetifyProps = (path: string) => {
    const props = get(appliedOptions.value?.vuetify, path);

    return props && isPlainObject(props) ? props : {};
  };

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    vuetifyProps,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = useJsonForms();

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};

export interface NestedInfo {
  level: number;
  parentElement?: 'array' | 'object';
}

export const useNested = (element: false | 'array' | 'object'): NestedInfo => {
  const nestedInfo = inject<NestedInfo>('jsonforms.nestedInfo', { level: 0 });
  if (element) {
    provide('jsonforms.nestedInfo', {
      level: nestedInfo.level + 1,
      parentElement: element,
    });
  }
  return nestedInfo;
};

export const useIcons = () => {
  const iconSet = computed<IconAliases>(() => {
    const icons = inject(IconSymbol);
    if (!icons) throw new Error('Missing Vuetify Icons provide!');

    let result = mdiIcons; // default
    const overrides = icons.aliases;

    if (icons.defaultSet === 'fa') {
      result = faIcons;
    }

    return overrides ? { ...result, ...overrides } : result;
  });

  return {
    current: iconSet,
  };
};
