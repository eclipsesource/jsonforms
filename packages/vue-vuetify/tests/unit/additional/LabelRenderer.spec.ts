import { describe, it, expect, beforeEach } from 'vitest';
import { clearAllIds } from '@jsonforms/core';
import LabelRenderer, {
  entry as labelRendererEntry,
} from '../../../src/additional/LabelRenderer.vue';
import { mountJsonForms } from '../util';

describe('LabelRenderer.vue', () => {
  const renderers = [labelRendererEntry];

  const data = '';
  const schema = {
    type: 'string',
  };
  const uischema = {
    type: 'Label',
    text: 'My Label',
  };

  let wrapper: ReturnType<typeof mountJsonForms>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  it('check if child LabelRenderer exists', () => {
    expect(wrapper.getComponent(LabelRenderer));
  });

  it('renders a label', () => {
    expect(wrapper.find('label').exists()).toBe(true);
  });

  it('renders label text', () => {
    expect(wrapper.find('label').text()).toEqual('My Label');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
