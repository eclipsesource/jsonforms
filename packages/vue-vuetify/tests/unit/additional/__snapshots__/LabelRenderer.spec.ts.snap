// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LabelRenderer.vue > should render component and match snapshot 1`] = `
"<div class="v-application v-theme--light v-layout v-layout--full-height v-locale--is-ltr">
  <div class="v-application__wrap"><label class="v-label label-element">
      <!---->My Label
    </label></div>
</div>"
`;
